### ISSUE TITLE
RandomCrop does nothing at validation/inference

### ISSUE THREAD

**Issue by ma7555** [keras-team/keras#21868]
The RandomCrop layer does absolutely nothing during validation. This is due to the if condition placed here:
 https://github.com/keras-team/keras/blob/adbfd13426a0da9864d9a0fcd5be5eed74ca341f/keras/src/layers/preprocessing/image_preprocessing/random_crop.py#L125

This looks pretty odd as the validation split will have different shape from the training data. The logic seemed to be it should do center crop for `training=False` as it is calculated here:
https://github.com/keras-team/keras/blob/adbfd13426a0da9864d9a0fcd5be5eed74ca341f/keras/src/layers/preprocessing/image_preprocessing/random_crop.py#L114

Example of training with 512x512 images but doing random crop to 256x512 confirms this as it raises this error as soon as the epoch finishes.
```
Input 0 of layer "convnext_small" is incompatible with the layer: expected shape=(None, 256, 512, 3), found shape=(None, 512, 512, 3)

Arguments received by Functional.call():
  • inputs=tf.Tensor(shape=(None, 512, 512, 3), dtype=float32)
  • training=False
  • mask=None
  • kwargs=<class 'inspect._empty'>
```

Please confirm if this is an intended behaviour that you must place a center crop layer with the same crop size just after it to maintain the pipeline?

**Comment by dhantule** [comment]
Hi @ma7555, thanks for reporting this. 

Can you provide code to reproduce this issue ?

**Comment by ma7555** [comment]
```
import matplotlib.pyplot as plt
import keras
import numpy as np

url = "https://storage.googleapis.com/download.tensorflow.org/example_images/grace_hopper.jpg"
image_path = keras.utils.get_file("grace_hopper.jpg", origin=url)

img = keras.utils.load_img(image_path, target_size=(512, 512))
img_array = keras.utils.img_to_array(img)
random_crop_layer = keras.layers.RandomCrop(height=256, width=256)

# Expectation: Different random crops (sliding window)
out1_random = random_crop_layer(img_array, training=True).numpy().astype('uint8')
out2_random = random_crop_layer(img_array, training=True).numpy().astype('uint8')

# Expectation: Identical Center Crops
out1_val = random_crop_layer(img_array, training=False).numpy().astype('uint8')
out2_val = random_crop_layer(img_array, training=False).numpy().astype('uint8')

fig, axes = plt.subplots(2, 2, figsize=(12, 8))

# Row 1: Training (Random)
axes[0, 0].imshow(out1_random)
axes[0, 0].set_title("Training=True (Random 1)")
axes[0, 1].imshow(out2_random)
axes[0, 1].set_title("Training=True (Random 2)")

# Row 2: Validation (Deterministic)
axes[1, 0].imshow(out1_val)
axes[1, 0].set_title("Training=False (Val 1 - No effect)")
axes[1, 1].imshow(out2_val)
axes[1, 1].set_title("Training=False (Val 2 - No effect)")

plt.tight_layout()
plt.savefig('randomcrop.png')
plt.show()
```

<img width="468" height="395" alt="Image" src="https://github.com/user-attachments/assets/2cb6c099-5df2-47c0-a391-d97c1aed0f9f" />

=======================================================================

Original image:

<img width="430" height="418" alt="Image" src="https://github.com/user-attachments/assets/35a1d7fe-9e1a-4f15-bcdb-de968713123b" />

**Comment by ma7555** [comment]
There is also a second bug that happens due to: https://github.com/keras-team/keras/blob/adbfd13426a0da9864d9a0fcd5be5eed74ca341f/keras/src/layers/preprocessing/image_preprocessing/random_crop.py#L95

The correct condition should be:
```
if training and input_height >= self.height and input_width >= self.width:
```
Because, as in my case, i am having a 512x512 images and doing random cropping to 256x512. The current condition will fail and will execute the else block which does center crop. Random crop should still work over one of the axis (as a sliding window).

**Comment by ma7555** [comment]
@utsab345 @dhantule 

One more thing to add, sometimes you may need to resize the validation instead of center crop. Although center crop is more common it may still be beneficial to allow resizing of validation images maybe through an argument that can accept either: 'resize' or 'center_crop' with a default of center_crop

What do you think?

**Comment by divyashreepathihalli** [comment]
@ma7555 - This is intended behavior. Generally it does not make sense to use augmentation layers on validation data during standard training. The goal of the validation set is to act as a stable "ruler" to measure your model's performance after every epoch. If you introduce Random Crop the Metric becomes noisy and the results will not be reproducible.

**Comment by ma7555** [comment]
Intended behavior to do nothing at validation? or do the center crop?
The debate here is whether to do center_crop or resizing during validation, otherwise you will get a shape mismatch error.

**Comment by hertschuh** [comment]
Reopening to discuss the fixed output size vs. not

**Comment by mattdangerw** [comment]
I'm somewhat inclined to leave the behavior as is, as I don't see an obvious "always right" solution for what to do at inference time. You might want to...

1. Run inference at the full image size despite training at a smaller size, for models that don't have a fixed input size requirement (most convnets). E.g. resnet models are frequently tested with images at [a slightly larger scale](https://huggingface.co/timm/resnet50.a1_in1k/blob/main/config.json#L8-L17) than training.
2. Resize the model to the target size without cropping for inference (with any of the arguments to the resizing layer).
3. Center crop the image.

I'd actually suspect that most users would probably want `2.` here, since throwing away image during training can make sense to augment and stretch your data, but will almost always result in loss of quality at inference time vs a resizing op that does not throw out portions of the image.

So it's probably better to ask users to be explicit here. Either only apply the `RandomCrop` during training, and apply `Resizing` or `CenterCrop` during inference. Or if you really want to use the same preprocessing pipeline at training and inference, simply chain the `RandomCrop` into a `Resizing` or `CenterCrop` layer (where the latter will be a no-op during training, since images will already be the target size, and only specifies the inference behavior).

Does that make sense @ma7555 ? If the center cropping during inference was going to be the right thing for 95% of users I'd be onboard with your proposal, but I don't think that's the case here. Most users should resize at inference.

I think it would be very reasonable to add a `apply_during_inference` flag to all the random layers (so they still do there thing at inference time). But a random crop should be a random crop always if it's appled. And use center crop if you want a center crop.

**Comment by Samir-atra** [comment]
It is just like what Divya explained which is the fact that the behavior is intentional and a feature and not a bug.

Because validation and inference is meant to have the real life data distribution and require to keep the full version of original image to make the best judgement possible and it has nothing to do with generalization or making the data distribution more feature rich and have wider coverage for the edge cases.

validation and more importantly inference is all about real life cases and complete reality. and editing the image means noise and inaccuracy.
